src/colorizer/UrlUtility.ts

```diff
@@ -2,68 +2,328 @@ const URL_PARAM_TRACK = "track";
-  }
-
-  /**
-   * Loads parameters from the current window URL.
-   * @returns An object with a dataset, feature, track, and time parameters.
-   * The dataset and feature parameters are null if no parameter was found in the URL, and the
-   * track and time will have negative values (-1) if no parameter (or an invalid parameter) was found.
-   */
-  public static loadParamsFromUrl(): { dataset: string | null; feature: string | null; track: number; time: number } {
-    // Get params from URL and load, with default fallbacks.
-    const queryString = window.location.search;
-    const urlParams = new URLSearchParams(queryString);
-
-    const base10Radix = 10; // required for parseInt
-    const datasetParam = urlParams.get(URL_PARAM_DATASET);
-    const featureParam = urlParams.get(URL_PARAM_FEATURE);
-    const trackParam = parseInt(urlParams.get(URL_PARAM_TRACK) || "-1", base10Radix);
-    // This assumes there are no negative timestamps in the dataset
-    const timeParam = parseInt(urlParams.get(URL_PARAM_TIME) || "-1", base10Radix);
-
-    return {
-      dataset: datasetParam,
-      feature: featureParam,
-      track: trackParam,
-      time: timeParam,
-    };
+export const DEFAULT_COLLECTION_PATH = "http://dev-aics-dtp-001.corp.alleninstitute.org/dan-data/colorizer/data";
+/**
+ * Default name for a collection descriptor JSON, which provides relative
+ * filepaths to one or more datasets.
+ */
+export const DEFAULT_COLLECTION_FILENAME = "collection.json";
+
+/**
+ * Default name for the manifest JSON, which provides relative filepaths
+ * to the elements (frames, feature data, centroids, etc.) of this dataset.
+ */
+export const DEFAULT_DATASET_FILENAME = "manifest.json";
+
+type UrlParams = {
+  collection: string | null;
+  dataset: string | null;
+  feature: string | null;
+  track: number;
+  time: number;
+};
+
+/**
+ * Dataset properties in a collection. Collections are defined as .json files containing an array of objects
+ * with the following properties:
+ *
+ * - `path`: a relative path from the base directory of the collection URL, or a URL to a dataset.
+ * - `name`: the display name for the dataset.
+ */
+export type CollectionEntry = {
+  path: string;
+  name: string;
+};
+
+export const DEFAULT_FETCH_TIMEOUT_MS = 5000;
+
+/**
+ * Initiates a fetch request with a given timeout, returning a promise that will reject if the timeout is reached.
+ * @param url fetch request URL
+ * @param timeoutMs timeout before the request should fail, in milliseconds. Defaults to `DEFAULT_FETCH_TIMEOUT_MS`.
+ * @param options additional
+ * @returns a Response promise, as returned by `fetch(url, options)`. The promise will reject if the timeout is exceeded.
+ */
+export function fetchWithTimeout(
+  url: string,
+  timeoutMs: number = DEFAULT_FETCH_TIMEOUT_MS,
+  options?: Object
+): Promise<Response> {
+  const controller = new AbortController();
+  const signal = controller.signal;
+  setTimeout(() => controller.abort, timeoutMs);
```

Comment:
what happens after timeoutMs when the fetch has already completed?

Reply:
So, the abort signal tells the fetch to reject its promise, but since the promise is already settled (resolved) it won't change its state. I think there's a test in RequestQueue over in volume-viewer that confirms that promise behavior.

I will, however, update this code so that it clears the timeout when the fetch completes, and I'll leave a comment about it.